src/inmanta/server/diff.py

```diff
@@ -0,0 +1,163 @@
+
+        self._compare_value: Optional[str] = None
+
+    @property
+    def value(self) -> object:
+        return self._value
+
+    @property
+    def compare_value(self) -> Optional[str]:
+        self._generate_compare_value()
+
+        return self._compare_value
+
+    def _generate_compare_value(self) -> None:
+        """Generate a value that can be used for comparison"""
+        if self._compare_value is not None:
+            return
+
+        if isinstance(self._value, list) and len(self._value) > 0 and not isinstance(self._value[0], (dict, list)):
+            # sort list that are not a json entry
+            self._compare_value = "\n".join([str(x) for x in sorted(self._value)])
+
+        elif isinstance(self._value, (dict, list)):  # json
+            self._compare_value = json.dumps(self._value, indent=4, sort_keys=True)
+        else:
+            self._compare_value = str(self.value)
+
+    def compare(self, other: "Attribute") -> Optional[AttributeDiff]:
+        """Compare this value with other. Other is considered the original value"""
+        if self.compare_value == other.compare_value:
+            return None
+
+        diff = AttributeDiff(
+            from_value=other.value,
+            to_value=self.value,
+            from_value_compare=other.compare_value,
+            to_value_compare=self.compare_value,
+        )
+        return diff
+
+    def added(self) -> AttributeDiff:
+        """Return an attribute diff as if this attribute is newly added"""
+        return AttributeDiff(
+            from_value=None,
+            to_value=self.value,
+            from_value_compare=None,
+            to_value_compare=self.compare_value,
+        )
+
+    def removed(self) -> AttributeDiff:
+        """Return an attribute diff as if this attribute is removed in a later version"""
+        return AttributeDiff(
+            from_value=self.value,
+            to_value=None,
+            from_value_compare=self.compare_value,
+            to_value_compare=None,
+        )
+
+
+class Resource:
+    def __init__(self, resource_id: ResourceIdStr, attributes: Dict[str, object]) -> None:
+        self.resource_id = resource_id
+        self._attributes = {name: Attribute(name, value) for name, value in attributes.items() if name != "version"}
+
+    def compare(self, other: "Resource") -> Optional[ResourceDiff]:
+        other_attributes = set(other._attributes.keys())
+        our_attributes = set(self._attributes.keys())
+
+        diff = {}
+        # Generate the diff for each attribute
+        for added in our_attributes - other_attributes:
+            diff[added] = self._attributes[added].added()
+
+        for removed in other_attributes - our_attributes:
+            diff[removed] = other._attributes[removed].removed()
+
+        for name in other_attributes.intersection(our_attributes):
+            attr_diff = self._attributes[name].compare(other._attributes[name])
```

Comment:
I would prefer the more symmetrical solution of implementing __sub__ and __rsub__ on `Attribute` but that is probably just me. 

```
for k in other_attribute + our_attributes:
   attr_diff = self._attributes.get(name) - other._attributes.get(name)
   if attr_diff is not None:
                diff[name] = attr_diff
```

so this is just a comment, no need to change it.